### PR TITLE
Enhanced comparison support

### PR DESCRIPTION
**Pull request by WillBrindle**
### GreaterThan

Implement GreaterThan action. Exact same as the LessThan action except, well, the other way round. Added regression test for this:
```
// Before introduction of ActionGreaterThan, the following if statement will be flipped around
// and ActionLessThan used, so nothing will be printed due to the side effects order. This test
// case is to confirm that swfs > 6 are using the GreaterThan operator and not flipped LessThan.
// When swf v < 6 expecting 'fail', but version >= 6 would expect 'success' to be printed.
var a = 0;
var b = 2;
function sideEffect() {
  a = 1;
  b = 0;

  return a;
}

if (sideEffect() > b) {
  trace("success");
} else {
  trace("fail");
}
```
This test is more complicated than needed, I mostly just wanted to have a case that behaves differently in fp5 and fp6. Added both swfs as tests of both cases.

### StrictEquals

Added strict equality. Same as equality but types must match. Added test cases:
```
if (2 === "2") {
	trace("ERROR: 2 === \"2\"");
}

if (2 === 2) {
	trace("2 === 2");
}

if (2 === 5) {
	trace("ERROR: 2 === 5");
}

if (true === true) {
	trace("true === true");
}

if (false === false) {
	trace("false === false");
}

if (true === false) {
	trace("ERROR: true === false");
}

if (1 === true) {
	trace("ERROR: 1 === true");
}

if (0 === false) {
	trace("ERROR: 0 === false");
}

if ("abc" === "abc") {
	trace("\"abc\" === \"abc\"");
}

if (0 === undefined) {
	trace("ERROR: 0 === undefined");
}

if (undefined === undefined) {
	trace("undefined === undefined");
}

if (NaN === NaN) {
	trace("NaN === NaN");
}

if (undefined === NaN) {
	trace("ERROR: undefined === NaN");
}

if (0 === null) {
	trace("ERROR: 0 === null");
}

if (null === null) {
	trace("null === null");
}

if (undefined === null) {
	trace("ERROR: undefined === null");
}

if (NaN === null) {
	trace("ERROR: NaN === null");
}
```
For NaN equality see below.

### StringGreater

Same as the GreaterThan action.

### NaN Equality?

A bit confused by this... I'm pretty sure that the spec says that `NaN == NaN` should be false... However testing this in flash I'm getting it back as true.  `trace (NaN == Nan)` is coming out as true. According to the swf19 spec:
>The equality comparison algorithm from ECMA-262 Section 11.9.3 is applied.

From the [ECMA-262 spec](https://www.ecma-international.org/ecma-262/5.1/#sec-11.9.3):
image

So my observation in flash seems to be at odds with this...

### CODE REVIEW

**Comment by wizzwizz4** [comment]
`NaN` shouldn't equal `NaN`.

**Comment by WillBrindle** [comment]
@wizzwizz4 that's what I thought from the spec. However, when testing in flash I'm seeing `if (NaN === NaN) { trace("NaN === NaN"); }` printed out. Anyway, I accidentally submitted this prematurely so need to check this further.

**Comment by wizzwizz4** [comment]
@WillBrindle That's confusing and bad.

File: core/src/avm1.rs [hunk 1]
```diff
@@ -1503,6 +1518,24 @@ impl<'gc> Avm1<'gc> {
...
+        let b = self.pop()?;
+        let result = match (b, a) {
+            (Value::Undefined, Value::Undefined) => true,
+            (Value::Null, Value::Null) => true,
+            (Value::Number(a), Value::Number(b)) => a == b || (a.is_nan() && b.is_nan()),
+            (Value::String(a), Value::String(b)) => a == b,
+            (Value::Bool(a), Value::Bool(b)) => a == b,
+            (Value::Object(_a), Value::Object(_b)) => false, // TODO
+            _ => false,
+        };
```
Comment (Dinnerbone) [inline core/src/avm1.rs]:
```suggestion
let result = a == b;
```

`Value` implements `PartialEq` which, as far as I can tell, is an identity equality check. It even implements for Object.

Comment (WillBrindle) [reply core/src/avm1.rs]: @Dinnerbone you're absolutely correct. Just checked and the 2 implementations are semantically the same. Thank you :)

**Comment by Herschel** [comment]
From the discussion in Discord, it looks like `NaN` is our first occurrence of Flash-Player-specific behavior 🎉

For the same v5 SWF file:
Flash Player <7 will output `NaN == NaN` as false.
Flash Player >=8 will output `NaN == NaN` as true.

Specifically Flash Player >= 8 does some wacky things:
```
var a = NaN;
trace(NaN === NaN);
trace(Number.NaN === Number.NaN);
trace(a === NaN);
trace(Number.NaN === a);
trace(a === a);
trace(Number(a) === Number(a));
```
will output a mixture of true and false on Flash Player 8+ (unexpected), but will return only false on Flash Player 7 and earlier (as expected). It's not totally clear to me how FP8 implements this (maybe NaN is considered as an object in some circumstances and compared reference-wise..?)

So basically we can do what we want here, as SWFs that rely on it will break in one situation on another. Ideally we behave completely like one version or the other. Long term, maybe there can be a dynamic switch we supply to the player to choose the behavior.

@WillBrindle, could you add a comment to the NaN == NaN tests noting that this behavior varies with player version?

**Comment by Herschel** [comment]
(BTW, Flash Player 4 `NaN == NaN` uses the older `ActionEquals` opcode which will return 1, but not relevant to the above because any newer version should use `ActionEquals2`.)
